### PR TITLE
extended scc handler

### PR DESCRIPTION
**Pull request by elenagerman**
new methods added to the scc package:
              - WithHostDirVolumePlugin
              - WithFSGroupRange
              - WithHostIPC
              - WithHostNetwork
              - WithHostPID
              - WithHostPorts
              - WithReadOnlyRootFilesystem
              - WithDefaultAddCapabilities
              - WithPriority
              - WithGroups
              - WithVolumes

### CODE REVIEW

File: pkg/scc/scc.go [hunk 1]
```diff
@@ -184,24 +268,94 @@ func (builder *Builder) WithAllowCapabilities(allowCapabilities []coreV1.Capabil
...
 
+// WithDefaultAddCapabilities adds list of defaultAddCapabilities to SecurityContextConstraints.
+func (builder *Builder) WithDefaultAddCapabilities(defaultAddCapabilities []coreV1.Capability) *Builder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof("Redefining SecurityContextConstraints %s with"+
+		"defaultAddCapabilities: %v", builder.Definition.Name, defaultAddCapabilities)
+
```
Comment (kononovn) [inline pkg/scc/scc.go:279]: Please add validation step: if len(allowCapabilities) == 0 like in functions above.

File: pkg/scc/scc.go [hunk 2]
```diff
@@ -184,24 +268,94 @@ func (builder *Builder) WithAllowCapabilities(allowCapabilities []coreV1.Capabil
...
+		"priority: %v", builder.Definition.Name, priority)
+
+	builder.Definition.Priority = priority
+
+	return builder
+}
+
 // WithFSGroup adds fsGroup to SecurityContextConstraints.
-func (builder *Builder) WithFSGroup(fsGroup string) *Builder {
+func (builder *Builder) WithFSGroup(fsGroupType string, fsGroupRange map[string]int64) *Builder {
```
Comment (kononovn) [inline pkg/scc/scc.go:307]: You are going to change public api functions WithFSGroup. This function is in use in few test cases. Let's avoid changing public api functions. 
Instead it's better to add additional WithFSGroupRange functions.

File: pkg/scc/scc.go [hunk 3]
```diff
@@ -184,24 +268,94 @@ func (builder *Builder) WithAllowCapabilities(allowCapabilities []coreV1.Capabil
...
-		builder.errorMsg = "SecurityContextConstraints 'fsGroup' cannot be empty string"
+		builder.errorMsg = "SecurityContextConstraints 'fsGroupType' cannot be empty string"
 
 		return builder
 	}
 
-	builder.Definition.FSGroup.Type = securityV1.FSGroupStrategyType(fsGroup)
+	builder.Definition.FSGroup.Type = securityV1.FSGroupStrategyType(fsGroupType)
+
+	if fsGroupRange == nil {
```
Comment (kononovn) [inline pkg/scc/scc.go:325]: In golang we can't verify is map is empty by this condition. Please use if len(fsGroupRange) == 0 instead.

File: pkg/scc/scc.go [hunk 4]
```diff
@@ -127,6 +127,90 @@ func (builder *Builder) WithPrivilegedEscalation(allowPrivilegedEscalation bool)
...
+// WithReadOnlyRootFilesystem adds bool flag to the readOnlyRootFilesystem of SecurityContextConstraints.
+func (builder *Builder) WithReadOnlyRootFilesystem(readOnlyRootFilesystem bool) *Builder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof("Redefining SecurityContextConstraints %s with"+
+		" readOnlyRootFilesystem: %t flag", builder.Definition.Name, readOnlyRootFilesystem)
+
+	builder.Definition.AllowHostPorts = readOnlyRootFilesystem
```
[the remark this answers is not in the document]
Comment (elenagerman) [reply pkg/scc/scc.go:209]: correct, thank you!

File: pkg/scc/scc.go (hunk 2 above)
Comment (kononovn) [inline pkg/scc/scc.go:307]: Let's avoid using complex self created structures and use just 2 argument's instead. fsGroupMin and fsGroupMax int64

File: pkg/scc/scc.go [hunk 5]
```diff
@@ -184,24 +268,94 @@ func (builder *Builder) WithAllowCapabilities(allowCapabilities []coreV1.Capabil
...
+	return builder
+}
+
+// WithGroups adds groups to SecurityContextConstraints.
+func (builder *Builder) WithGroups(groups []string) *Builder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof("Redefining SecurityContextConstraints %s with"+
```
Comment (kononovn) [inline pkg/scc/scc.go:347]: Please change "groups: %s" to "groups: %v"

File: pkg/scc/scc.go [hunk 6]
```diff
@@ -184,24 +268,94 @@ func (builder *Builder) WithAllowCapabilities(allowCapabilities []coreV1.Capabil
...
+		"groups: %s", builder.Definition.Name, groups)
```
Comment (kononovn) [inline pkg/scc/scc.go:348]: Please change "groups: %s" to "groups: %v"

File: pkg/scc/scc.go [hunk 7]
```diff
@@ -184,24 +268,94 @@ func (builder *Builder) WithAllowCapabilities(allowCapabilities []coreV1.Capabil
...
+
+	if groups == nil {
```
Comment (kononovn) [inline pkg/scc/scc.go:350]: Please replace validation by if len(groups) == 0

File: pkg/scc/scc.go [hunk 8]
```diff
@@ -184,24 +268,94 @@ func (builder *Builder) WithAllowCapabilities(allowCapabilities []coreV1.Capabil
...
+		glog.V(100).Infof("SecurityContextConstraints 'groups' argument cannot be empty")
+
+		builder.errorMsg = "SecurityContextConstraints 'fsGroupType' cannot be empty string"
+
+		return builder
+	}
+
+	builder.Definition.Groups = groups
```
Comment (kononovn) [inline pkg/scc/scc.go:358]: Let's make this function a bit more generic.
if builder.Definition.Groups != nil{
use append 
}

File: pkg/scc/scc.go [hunk 9]
```diff
@@ -284,6 +438,26 @@ func (builder *Builder) WithUsers(users []string) *Builder {
...
+// WithVolumes adds list of volumes to SecurityContextConstraints.
+func (builder *Builder) WithVolumes(volumes []securityV1.FSType) *Builder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof("Redefining SecurityContextConstraints %s with"+
+		"volumes: %v", builder.Definition.Name, volumes)
+
+	if builder.Definition.Volumes == nil {
```
Comment (kononovn) [inline pkg/scc/scc.go:450]: Please use len validation instead.

File: pkg/scc/scc.go (hunk 2 above)
Comment (elenagerman) [reply pkg/scc/scc.go:307]: I don't have a problem to change it back, just take in mind that parameter used as fsGroup is fsGroupType, so there is some inconsistency.

**Review by kononovn — CHANGES_REQUESTED** [review]
Just few minor comments to fix

File: pkg/scc/scc.go (hunk 9 above)
Comment (kononovn) [reply pkg/scc/scc.go]: Sorry, i meant volumes argument and not builder.Definition.Volumes.
Please check function below:
```

// WithVolumes adds list of volumes to SecurityContextConstraints.
func (builder *Builder) WithVolumes(volumes []securityV1.FSType) *Builder {
	if valid, _ := builder.validate(); !valid {
		return builder
	}

	glog.V(100).Infof("Redefining SecurityContextConstraints %s with"+
		"volumes: %v", builder.Definition.Name, volumes)

	if len(volumes) == 0 {
		glog.V(100).Infof("SecurityContextConstraints 'volumes' argument cannot be empty")

		builder.errorMsg = "SecurityContextConstraints 'volumes' cannot be empty list"
		
		return builder
	}
	
	if builder.Definition.Volumes == nil{
		builder.Definition.Volumes = volumes
		
		return builder
	}

	builder.Definition.Volumes = append(builder.Definition.Volumes, volumes...)
	
	return builder
}
```

File: pkg/scc/scc.go [hunk 10]
```diff
@@ -203,6 +330,45 @@ func (builder *Builder) WithFSGroup(fsGroup string) *Builder {
...
+	if len(groups) == 0 {
+		glog.V(100).Infof("SecurityContextConstraints 'groups' argument cannot be empty")
+
+		builder.errorMsg = "SecurityContextConstraints 'fsGroupType' cannot be empty string"
+
+		return builder
+	}
+
+	if builder.Definition.Groups == nil {
+		builder.Definition.Groups = []string{}
```
Comment (kononovn) [inline pkg/scc/scc.go:367]: Invalid expression. It should be :  
builder.Definition.Groups = groups

File: pkg/scc/scc.go (hunk 8 above)
Comment (kononovn) [reply pkg/scc/scc.go:370]: Please use append here:
builder.Definition.Groups = append(builder.Definition.Groups, groups...)

File: pkg/scc/scc.go [hunk 11]
```diff
@@ -203,6 +330,45 @@ func (builder *Builder) WithFSGroup(fsGroup string) *Builder {
...
+		"groups: %v", builder.Definition.Name, groups)
+
+	if len(groups) == 0 {
+		glog.V(100).Infof("SecurityContextConstraints 'groups' argument cannot be empty")
+
+		builder.errorMsg = "SecurityContextConstraints 'fsGroupType' cannot be empty string"
+
+		return builder
+	}
+
```
Comment (kononovn) [inline pkg/scc/scc.go:365]: Please check that groups is not empty list by
if len(groups) == 0

File: pkg/scc/scc.go (hunk 2 above)
Comment (kononovn) [reply pkg/scc/scc.go]: I suggested to add new function. Please check below:

```
func (builder *Builder) WithFSGroupRange(fsGroupMin, fsGroupMax int64) *Builder {
	if valid, _ := builder.validate(); !valid {
		return builder
	}

	glog.V(100).Infof("Redefining SecurityContextConstraints %s with"+
		"fsGroupRange: fsGroupMin: %d, fsGroupMax: %d", builder.Definition.Name, fsGroupMin, fsGroupMax)

	if fsGroupMin > fsGroupMax {
		glog.V(100).Infof("SecurityContextConstraints 'fsGroupMin' argument can not be greater than fsGroupMax")

		builder.errorMsg = "SecurityContextConstraints 'fsGroupMin' argument can not be greater than fsGroupMax"

		return builder
	}

	if builder.Definition.FSGroup.Ranges == nil {
		builder.Definition.FSGroup.Ranges = []securityV1.IDRange{{Min: fsGroupMin, Max: fsGroupMax}}

		return builder
	}

	builder.Definition.FSGroup.Ranges = append(
		builder.Definition.FSGroup.Ranges, securityV1.IDRange{Max: fsGroupMax, Min: fsGroupMin})

	return builder
}
```

**Review by kononovn — CHANGES_REQUESTED** [review]
Please check few comments above and few suggestions.

File: pkg/scc/scc.go (hunk 2 above)
Comment (kononovn) [reply pkg/scc/scc.go]: Please revert WithFSGroup to it's original state.

File: pkg/scc/scc.go (hunk 11 above)
Comment (kononovn) [reply pkg/scc/scc.go:380]: @elenagerman Can you please fix this comment ?

File: pkg/scc/scc.go [hunk 12]
```diff
@@ -284,6 +465,26 @@ func (builder *Builder) WithUsers(users []string) *Builder {
 	return builder
 }
 
+// WithVolumes adds list of volumes to SecurityContextConstraints.
+func (builder *Builder) WithVolumes(volumes []securityV1.FSType) *Builder {
```
Comment (kononovn) [inline pkg/scc/scc.go:469]: Can you please endit this functiton? volumes size validation is missing. Please take a look below:

// WithVolumes adds list of volumes to SecurityContextConstraints.
func (builder *Builder) WithVolumes(volumes []securityV1.FSType) *Builder {
	if valid, _ := builder.validate(); !valid {
		return builder
	}

	glog.V(100).Infof("Redefining SecurityContextConstraints %s with"+
		"volumes: %v", builder.Definition.Name, volumes)

	if len(volumes) == 0 {
		glog.V(100).Infof("SecurityContextConstraints 'volumes' argument cannot be empty")

		builder.errorMsg = "SecurityContextConstraints 'volumes' cannot be empty list"
		
		return builder
	}
	
	if builder.Definition.Volumes == nil{
		builder.Definition.Volumes = volumes
		
		return builder
	}

	builder.Definition.Volumes = append(builder.Definition.Volumes, volumes...)
	
	return builder
}

File: pkg/scc/scc.go [hunk 13]
```diff
@@ -206,6 +333,60 @@ func (builder *Builder) WithFSGroup(fsGroup string) *Builder {
...
+		glog.V(100).Infof("SecurityContextConstraints 'groups' argument cannot be empty")
+
+		builder.errorMsg = "SecurityContextConstraints 'fsGroupType' cannot be empty string"
+
+		return builder
+	}
+
+	if builder.Definition.Groups == nil {
+		builder.Definition.Groups = groups
+	} else {
```
Comment (kononovn) [inline pkg/scc/scc.go:383]: Inconsistent code else. In all other function we use different construction.

Please replace by:
```
if builder.Definition.Groups == nil {
		builder.Definition.Groups = groups

                return builder
}

builder.Definition.Groups = append(builder.Definition.Groups, groups...)

retrun builder 
```

File: pkg/scc/scc.go [hunk 14]
```diff
@@ -206,6 +333,60 @@ func (builder *Builder) WithFSGroup(fsGroup string) *Builder {
 	return builder
 }
 
+func (builder *Builder) WithFSGroupRange(fsGroupMin, fsGroupMax int64) *Builder {
```
Comment (kononovn) [inline pkg/scc/scc.go:336]: Please add function comment above

**Review by kononovn — CHANGES_REQUESTED** [review]
Just few more minor changed are required.

File: pkg/scc/scc.go (hunk 12 above)
Comment (elenagerman) [reply pkg/scc/scc.go:469]: np, only one question: last time you ask me to replace 
`if builder.Definition.Volumes == nil{`
with
`if len(builder.Definition.Volumes) == 0 {`
is it still relevant?

File: pkg/scc/scc.go (hunk 14 above)
Comment (elenagerman) [reply pkg/scc/scc.go:336]: oh, yes, sorry

File: pkg/scc/scc.go (hunk 12 above)
Comment (kononovn) [reply pkg/scc/scc.go:472]: I meant **volumes** argument not **builder.Definition.Volumes**. I just copy/pasted wrong string. 
In this comment https://github.com/openshift-kni/eco-goinfra/pull/129#discussion_r1336273234 I shared that I made a mistake.

Comment (kononovn) [reply pkg/scc/scc.go:472]: Please revert it back to **if builder.Definition.Volumes == nil** instead of **if len(builder.Definition.Volumes) == 0 {**

**Review by kononovn — CHANGES_REQUESTED** [review]
Please fix the last comment above

**Review by kononovn — APPROVED** [review]
/lgtm

**Review by trewest — APPROVED** [review]
/lgtm

**Comment by trewest** [comment]
@elenagerman looks good, can you just fix the typo in the commit message - extended scc handler

**Pull request merged**